Mods/src/PuppiMod.cc

```diff
@@ -385,69 +419,106 @@ PuppiMod::Process()
+        PuppiParticle->SetPtEtaPhiM(PuppiParticle->Pt() * weight,PuppiParticle->Eta(),
+                                    PuppiParticle->Phi(),PuppiParticle->Mass() * weight);
+    }
+    if (fBothPVandPU) {
+      weight = 1.0 - weight;
+      if (weight == 0)
+        continue;
+      PFCandidate *PuppiParticle = InvertedParticles->Allocate();
+      new (PuppiParticle) PFCandidate(*pfCandidates->At(iCand));
+      if (weight < 1)                                                    // Weight the particle if required
+        PuppiParticle->SetPtEtaPhiM(PuppiParticle->Pt() * weight,PuppiParticle->Eta(),
+                                    PuppiParticle->Phi(),PuppiParticle->Mass() * weight);
+    }
+  }
+
+  if (fNoLepton) {                                                    // If no leptons were used, add them back in
+    for (UInt_t iLepton = 0; iLepton < leptonCandidates->GetEntries(); iLepton++) {
+      auto* cand = leptonCandidates->At(iLepton);
+      if (GetParticleType(cand, pv) != kChargedPU) {                   // From PV, simply add with weight 1
+        PFCandidate *PuppiParticle = fPuppiParticles->Allocate();
+        new (PuppiParticle) PFCandidate(*cand);
       }
-      std::vector<UInt_t> nearList(fMapper->GetSurrounding(iCand));
-      std::cout << "Nearby PV particles: ";
-      for (UInt_t iNeighbor : nearList) {
-        if (iNeighbor == iCand)
-          continue;
-
-        if (GetParticleType(pfCandidates->At(iNeighbor), pv) == kChargedPrimary)
-          std::cout << iNeighbor << " (" << MathUtils::DeltaR(PuppiParticle->Eta(),PuppiParticle->Phi(),
-                                                                 pfCandidates->At(iNeighbor)->Eta(),
-                                                                 pfCandidates->At(iNeighbor)->Phi()) <<  "), ";
+      else if (fBothPVandPU) {                                         // If PU, but using Inverted Puppi, add to inverted
+        PFCandidate *PuppiParticle = InvertedParticles->Allocate();
+        new (PuppiParticle) PFCandidate(*cand);
       }
-      std::cout << std::endl;
-      std::cout << "Nearby other particles: ";
-      for (UInt_t iNeighbor : nearList) {
-        if (iNeighbor == iCand)
-          continue;
-
-        if (GetParticleType(pfCandidates->At(iNeighbor), pv) != kChargedPrimary)
-          std::cout << iNeighbor << " (" << MathUtils::DeltaR(PuppiParticle->Eta(),PuppiParticle->Phi(),
-                                                                 pfCandidates->At(iNeighbor)->Eta(),
-                                                                 pfCandidates->At(iNeighbor)->Phi()) <<  "), ";
+      else if (fKeepPileup) {                                          // If keeping pileup, just add with weight 0
+        PFCandidate *PuppiParticle = fPuppiParticles->Allocate();
+        new (PuppiParticle) PFCandidate(*cand);
+        Double_t weight = 0;
```

Comment:
Not a big deal but you are overriding the variable weight here. Rather than declaring a variable that has value 0 and multiplying function calls with that variable, perhaps you should just pass 0 for pT and mass of the lepton?